### PR TITLE
arangodump: Shard-parallelism and smaller dumps

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

* Add shard-parallelism to arangodump when dumping collections with multiple shards.
  Previously, arangodump could execute a dump concurrently on different collections, but it did not parallelize the dump for multiple shards of the same collection.
  This change should speed up dumping of collections with multiple shards. 
  When dumping multiple shards of the same collection concurrently, Parallelism is still limited by all these threads needing to serialize their chunks into the same (shared) output file. 

* Add option `--envelope` to arangodump, to control if each dumped document should be wrapped into a small JSON wrapper (e.g. `{"type":2300,"data":{...}}`). 
  This JSON wrapper is not necessary anymore since ArangoDB 3.8, so omitting it produces smaller (and slightly faster) dumps.
  Restoring a dump without these JSON wrappers is handled automatically by ArangoDB 3.8 and higher. However, restoring a dump without these JSON wrappers into previous versions (pre v3.8) is not supported. Thus the option should only be used if the client tools (arangodump, arangorestore) and the arangod server are all using v3.8 or higher.
  The default value for this option is `true`, meaning the JSON wrappers will be stored as part of the dump. This is compatible with all previous verions.

Oskar PR: https://github.com/arangodb/oskar/pull/287
Blocklist PR for 3.6: https://github.com/arangodb/arangodb/pull/13402
Blocklist PR for 3.7: https://github.com/arangodb/arangodb/pull/13403

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/610

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *dump, dump_multiple, dump_encrypted*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new **integration tests** (e.g. in dump_no_envelope)

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13649/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13650/

**Review by neunhoef — APPROVED** [review]
LGTM

**Comment by jsteemann** [comment]
Tests blue

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 1]
```diff
@@ -1413,106 +1404,129 @@ Result RestReplicationHandler::processRestoreData(std::string const& colName) {
...
+  VPackBuilder documentsToInsert;
 
-  Result res = parseBatch(StaticStrings::AnalyzersCollection, latest, allMarkers);
-  if (res.fail()) {
-    return res;
-  }
-  VPackSlice allMarkersSlice = allMarkers.slice();
+  {
+    // this "fake" transaction here is only needed to get access to the underlying
+    // _users collection. we will not write anything into the collection here
```
Comment (mpoeter) [inline arangod/RestHandler/RestReplicationHandler.cpp:1511]:
```suggestion
    // _analyzers collection. we will not write anything into the collection here
```

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 2]
```diff
@@ -1522,16 +1536,24 @@ Result RestReplicationHandler::processRestoreCoordinatorAnalyzersBatch() {
...
+    SingleCollectionTransaction trx(ctx, StaticStrings::UsersCollection, AccessMode::Type::READ);
 
-  VPackSlice allMarkersSlice = allMarkers.slice();
+    Result res = trx.begin();
+    if (res.ok()) {
+      res = parseBatch(trx, StaticStrings::UsersCollection, documentsToInsert, generateNewRevisionIds);
+    }
+    if (res.fail()) {
+      return res;
+    }
```
Comment (mpoeter) [inline arangod/RestHandler/RestReplicationHandler.cpp:1554]: These lines are identical to those in `processRestoreCoordinatorAnalyzersBatch` (except for the collection name). Can we keep it DRY and extract them into a separate function?

**Review by mpoeter — APPROVED** [review]

File: arangod/RestHandler/RestReplicationHandler.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/RestHandler/RestReplicationHandler.cpp]: Good catch. Done.

File: arangod/RestHandler/RestReplicationHandler.cpp [hunk 3]
```diff
@@ -184,54 +184,45 @@ static Result restoreDataParser(char const* ptr, char const* pos,
...
-        "got document marker without object contents for collection '" + collectionName +
-            "' on line " + std::to_string(line) + ": " + doc.toJson()};
-  }
+  TRI_ASSERT(type != REPLICATION_MARKER_DOCUMENT || doc.isObject());
 
-  if (key.empty()) {
-    return Result{TRI_ERROR_HTTP_BAD_PARAMETER,
-                  "received invalid JSON data for collection '" + collectionName +
-                      "' on line " + std::to_string(line) + ": empty key"};
+  if (type == REPLICATION_INVALID) {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/RestHandler/RestReplicationHandler.cpp]: Done.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13665/

**Comment by jsteemann** [comment]
Tests blue

**Pull request merged**